### PR TITLE
TypeScript template

### LINKED ISSUE #232: Provide a template with TypeScript

### PR DESCRIPTION
**Pull request by keanulee**
PR opened for review comments only - this branch serves as a template on how to use TypeScript.

Fixes #232

### CODE REVIEW

**Comment by abdonrd** [comment]
It seems that this template does not work with the new Typescript 3.1.

screenshot 2018-10-05 at 00 30 27

**Comment by keanulee** [comment]
It's mostly finished. Currently waiting on:

- Next pwa-helpers release, where underscore is removed from `stateChanged()` to match LitElement's convention of no underscore for protected methods (https://github.com/Polymer/pwa-helpers/pull/43)
- Next LitElement release, where LitElement is no longer abstract so we don't need to do the "Connected" element subclassing thing

Haven't looked into the TS 3.1 issues - that's probably for another PR. The process of "releasing" this template will probably just be renaming this branch to `template-typescript`.

**Comment by abdonrd** [comment]
> * Next pwa-helpers release, where underscore is removed from `stateChanged()` to match LitElement's convention of no underscore for protected methods ([Polymer/pwa-helpers#43](https://github.com/Polymer/pwa-helpers/pull/43))

About the underscore, you are also going to eliminate it from the properties?

**Comment by keanulee** [comment]
@abdonrd no - the convention is no underscore on protected methods

**Comment by abdonrd** [comment]
By the way, any plan to add `watch` to the `npm start` script to recompile on changes?

**Comment by abdonrd** [comment]
Also, any plan to setup some TSLint config?

**Comment by keanulee** [comment]
watch would be nice-to-have for `npm start`; haven't got around to that yet. PR welcome.

**Comment by abdonrd** [comment]
> watch would be nice-to-have for `npm start`; haven't got around to that yet. PR welcome.

Done at #244!

**Comment by keanulee** [comment]
TS 3.1 issue - blocked on release of https://github.com/Polymer/polymer/pull/5370

**Comment by felipecesar42** [comment]
Sugestion: organize better the js files. The number of files of the directory is 4x bigges because of d.ts, js, js.map. Maybe it would be good to create a new directory just to keep the generated files. I think it also would be a good idea to put a watch flag in tsc so people don't need to run "npm start" often to see the changes
screen shot 2018-10-10 at 16 08 18

**Comment by mercmobily** [comment]
@felipecesar42  I agree with that. That's what lit-html does by setting `outDir` in `tsconfig.json`

**Comment by keanulee** [comment]
Ack the number of files comment. Could consider moving TS source to a `ts_src/` directory or similar, with the idea that `src/` still contains JS so that `polymer serve`/`polymer build` continues to work without further changes. Will leave that for another PR.

CLA ok

File: src/actions/app.ts [hunk 1]
```diff
@@ -8,13 +8,25 @@ Code distributed by Google as part of the polymer project is also
...
+import { ThunkAction } from 'redux-thunk';
+import { RootState } from '../store.js';
 export const UPDATE_PAGE = 'UPDATE_PAGE';
 export const UPDATE_OFFLINE = 'UPDATE_OFFLINE';
 export const UPDATE_DRAWER_STATE = 'UPDATE_DRAWER_STATE';
 export const OPEN_SNACKBAR = 'OPEN_SNACKBAR';
 export const CLOSE_SNACKBAR = 'CLOSE_SNACKBAR';
 
-export const navigate = (path) => (dispatch) => {
+export interface AppActionUpdatePage extends Action<'UPDATE_PAGE'> {page: string};
```
Comment (abdonrd) [inline src/actions/app.ts]: @keanulee why export these interfaces?

Comment (keanulee) [reply src/actions/app.ts]: Oh, I guess store.js only uses `AppAction`, not these interfaces. No good reason, but maybe someone might want to create the action before dispatching it?

```js
const action = actionCreator(param);
store.dispatch(action);
```

**Pull request merged**